extension/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+            elog!(Error, "input time less than lowest time")
+        }
+
+        let mut state = state.unwrap_or_else(|| {
+            let state: Pox<_> = GapfillDeltaTransition::new(
+                lowest_time,
+                greatest_time,
+                window_size,
+                step_size,
+                true,
+                true,
+            ).into();
+            state
+        });
+
+        state.add_data_point(time, val);
+
+        Some(state)
+    }
+
+    pub fn gapfill_increase_transition(
+        state: Option<Pox<GapfillDeltaTransition>>,
+        lowest_time: TimestampTz,
+        greatest_time: TimestampTz,
+        step_size: Milliseconds, // `prev_now - step` is where the next window starts
+        window_size: Milliseconds, // the size of a window to delta over
+        time: TimestampTz,
+        val: f64,
+    ) -> Option<Pox<GapfillDeltaTransition>> {
+        if time < lowest_time || time > greatest_time {
+            elog!(Error, "input time less than lowest time")
+        }
+
+        let mut state = state.unwrap_or_else(|| {
+            let state: Pox<_> = GapfillDeltaTransition::new(
+                lowest_time,
+                greatest_time,
+                window_size,
+                step_size,
+                true,
+                false,
+            ).into();
+            state
+        });
+
+        state.add_data_point(time, val);
+
+        Some(state)
+    }
+
+    pub fn gapfill_delta_final(state: Option<Pox<GapfillDeltaTransition>>) -> Option<*mut ArrayType> {
+        state.map(|mut s| s.to_pg_array())
+    }
+}
+
+#[derive(Debug)]
+struct GapfillDeltaTransition {
+    window: VecDeque<(TimestampTz, f64)>,
+    deltas: Vec<Datum>,
+    nulls: Vec<bool>,
+    current_window_max: TimestampTz,
+    current_window_min: TimestampTz,
+    step_size: TimestampTz,
+    greatest_time: TimestampTz,
+    is_counter: bool,
+    is_rate: bool,
+}
+
+impl GapfillDeltaTransition {
+    pub fn new(
+        lowest_time: TimestampTz,
+        greatest_time: TimestampTz,
+        window_size: Milliseconds,
+        step_size: Milliseconds,
+        is_counter: bool,
+        is_rate: bool
+    ) -> Self {
+        let mut expected_deltas = (greatest_time - lowest_time) / (step_size * USECS_PER_MS);
+        if (greatest_time - lowest_time) % (step_size * USECS_PER_MS) != 0 {
+            expected_deltas += 1
+        }
+        GapfillDeltaTransition{
+            window: VecDeque::default(),
+            deltas: Vec::with_capacity(expected_deltas as usize),
+            nulls: Vec::with_capacity(expected_deltas as usize),
+            current_window_max: lowest_time + window_size*USECS_PER_MS,
+            current_window_min: lowest_time,
+            step_size: step_size*USECS_PER_MS,
+            greatest_time,
+            is_counter,
+            is_rate,
+        }
+    }
+
+    fn add_data_point(&mut self, time: TimestampTz, val: f64) {
```

Comment:
do we check somewhere that data is coming in in the correct order wrt time?

Reply:
nope, do want me to?

Reply:
yes pls.